File: research/experiments/power-thermal-probes/iopowersources_probe.m
#import <Foundation/Foundation.h>
#import <IOKit/ps/IOPowerSources.h>
#import <IOKit/ps/IOPSKeys.h>

static NSString *ValueString(NSDictionary *dictionary, NSString *key) {
    id value = dictionary[key];
    return value ? [value description] : @"n/a";
}

static NSString *KeyString(const char *key) {
    return [NSString stringWithUTF8String:key];
}

int main(void) {
    @autoreleasepool {
        CFTypeRef blob = IOPSCopyPowerSourcesInfo();
        CFArrayRef sources = IOPSCopyPowerSourcesList(blob);
        CFIndex count = CFArrayGetCount(sources);
        CFTimeInterval estimate = IOPSGetTimeRemainingEstimate();

        printf("source_count=%ld\n", (long)count);
        if (estimate == kIOPSTimeRemainingUnlimited) {
            puts("estimated_runtime=unlimited");
        } else if (estimate == kIOPSTimeRemainingUnknown) {
            puts("estimated_runtime=unknown");
        } else {
            printf("estimated_runtime_seconds=%.0f\n", estimate);
        }

        for (CFIndex index = 0; index < count; index++) {
            CFTypeRef source = CFArrayGetValueAtIndex(sources, index);
            CFDictionaryRef raw = IOPSGetPowerSourceDescription(blob, source);
            NSDictionary *description = (__bridge NSDictionary *)raw;
            printf("source[%ld].name=%s\n", (long)index, ValueString(description, KeyString(kIOPSNameKey)).UTF8String);
            printf("source[%ld].state=%s\n", (long)index, ValueString(description, KeyString(kIOPSPowerSourceStateKey)).UTF8String);
            printf("source[%ld].present=%s\n", (long)index, ValueString(description, KeyString(kIOPSIsPresentKey)).UTF8String);
            printf("source[%ld].current_capacity=%s\n", (long)index, ValueString(description, KeyString(kIOPSCurrentCapacityKey)).UTF8String);
            printf("source[%ld].max_capacity=%s\n", (long)index, ValueString(description, KeyString(kIOPSMaxCapacityKey)).UTF8String);
            printf("source[%ld].is_charging=%s\n", (long)index, ValueString(description, KeyString(kIOPSIsChargingKey)).UTF8String);
            printf("source[%ld].is_charged=%s\n", (long)index, ValueString(description, KeyString(kIOPSIsChargedKey)).UTF8String);
            printf("source[%ld].current=%s\n", (long)index, ValueString(description, KeyString(kIOPSCurrentKey)).UTF8String);
            printf("source[%ld].voltage=%s\n", (long)index, ValueString(description, KeyString(kIOPSVoltageKey)).UTF8String);
            printf("source[%ld].time_to_empty=%s\n", (long)index, ValueString(description, KeyString(kIOPSTimeToEmptyKey)).UTF8String);
        }

        CFRelease(sources);
        CFRelease(blob);
    }
    return 0;
}

File: research/experiments/power-thermal-probes/iopowersources_probe.swift
import Foundation
import IOKit
import IOKit.ps

func number(_ value: Any?) -> String {
    if let value = value as? NSNumber {
        return value.stringValue
    }
    return "n/a"
}
let blob = IOPSCopyPowerSourcesInfo().takeRetainedValue()
let sources = IOPSCopyPowerSourcesList(blob).takeRetainedValue() as [CFTypeRef]
let estimate = IOPSGetTimeRemainingEstimate()

print("source_count=\(sources.count)")
if estimate == kIOPSTimeRemainingUnlimited {
    print("estimated_runtime=unlimited")
} else if estimate == kIOPSTimeRemainingUnknown {
    print("estimated_runtime=unknown")
} else {
    print("estimated_runtime_seconds=\(estimate)")
}

for (index, source) in sources.enumerated() {
    guard let description = IOPSGetPowerSourceDescription(blob, source)?.takeUnretainedValue() as? [String: Any] else {
        print("source[\(index)]=unavailable")
        continue
    }

    print("source[\(index)].name=\(description[kIOPSNameKey as String] ?? "n/a")")
    print("source[\(index)].state=\(description[kIOPSPowerSourceStateKey as String] ?? "n/a")")
    print("source[\(index)].present=\(description[kIOPSIsPresentKey as String] ?? "n/a")")
    print("source[\(index)].current_capacity=\(number(description[kIOPSCurrentCapacityKey as String]))")
    print("source[\(index)].max_capacity=\(number(description[kIOPSMaxCapacityKey as String]))")
    print("source[\(index)].is_charging=\(description[kIOPSIsChargingKey as String] ?? "n/a")")
    print("source[\(index)].is_charged=\(description[kIOPSIsChargedKey as String] ?? "n/a")")
    print("source[\(index)].current=\(number(description[kIOPSCurrentKey as String]))")
    print("source[\(index)].voltage=\(number(description[kIOPSVoltageKey as String]))")
    print("source[\(index)].time_to_empty=\(number(description[kIOPSTimeToEmptyKey as String]))")
}
